docs(defining-templates): restructure and improve defining templates section (#6553)
camunda/camunda-docs: docs/components/modeler/element-templates/template-properties.md

The underlying concept is very simple:

1. You will create one property object for each property that should be defined by the template.
2. Each property object contains key-value pairs to define how the property is presented and how its value can be changed by the user.
3. Each property object contains one `binding` object that specifies how the property is mapped to BPMN 2.0 XML.

Element templates can only set properties that can be described by bindings supported by the element template schema.
You can find the full list of supported bindings in the [bindings](#binding-an-input-to-a-bpmn-or-camunda-element-property-binding) section.

When the user applies a template, the properties panel hides all BPMN 2.0 XML and Camunda extension elements properties that can be defined by bindings in an element template.
The template author must explicitly make the properties user configurable to show them in the properties panel once the user has applied the template.
For example, if a template does not contain a property object with the binding type `zeebe:input`,
the template user will not be able to define an input mapping for the element once the template is applied.

:::info

> Some properties, such as execution listeners, task listeners, element documentation, and multi-instance configurations cannot be set by element templates.

Author: like so?

For more information, see the documentation below.
If your editor supports JSON schema, these incompatibilities or missing keys-value pairs are highlighted while you edit your template.
For most purposes, `binding`, `label`, `type`, and `value` are sufficient to define a property.

All property objects are defined inside the `properties` array:

```json
{
  "schema": "https://unpkg.com/@camunda/zeebe-element-templates-json-schema/resources/schema.json",
  ...,
  "properties": [
    {
      "label": "Some property",
      "type": "String",
      "binding": {
        ...
      }
    },
    {
      "label": "Some other property",
      "type": "Number",
      "binding": {
        ...
      }
    },
    ...
  ]
}
```

For a comprehensive example showing how to create a REST connector template with all the concepts covered in this documentation, see the [complete template example](./template-example.md) page.

The key-value pairs of the property object are explained in the following sections.

## Setting a default value: `value`

The `value` key defines a static default value for a property.

> The value is applied to the property when the template is applied to an element until a user provides their own input value.

Author: Not really, the value is set on the property defined by the binding

`value` should be defined whenever the input type is not `Hidden`.
The value of `value` must match the `type` of the property.

```json
{
  "value": "4",

>   "type": "Hidden",

Author: Yeah, that was a typo above :D. You should use a value when the input is hidden otherwise you are not defining anything

  "binding": {
    "type": "zeebe:taskDefinition",
    "property": "retries"
  }
}
```

## Generating a value: `generatedValue`

As an alternative to static `value`, you can use a generated value. The value is generated when a property is applied to an element. Currently, the generated value can be a UUID:

```json
{
  "generatedValue": {
    "type": "uuid"
  },
  "type": "Hidden",
  "binding": {
    "type": "zeebe:property",
    "name": "id"
  }
}
```

## Setting a text placeholder: `placeholder`

The following property types support the `placeholder` attribute:

> - `String`

Author: Yeah, there is really no perfect order here. we could move placeholder down. Wdyt?

}
```

## Setting the input type: `type`

The input types `String`, `Text`, `Number`, `Boolean`, `Dropdown`, and `Hidden` are available.

### String input type

The `String` type maps to a single-line input field. Refer to the [FEEL](#adding-feel-editor-support-feel) section to use `Strings` as an expression.

### Text input type

The `Text` type maps to a multi-line text area. Refer to the [FEEL](#adding-feel-editor-support-feel) section to use `Text` as an expression.

### Hidden input type

The `Hidden` type is not shown in the properties panel. It is used to set [static](#setting-a-default-value-value) or [generated](#generating-a-value-generatedvalue) values that should not be changed by the user.

### Number input type

The `Number` type maps to a number input field. Refer to the [FEEL](#adding-feel-editor-support-feel) section to use `Numbers` as expressions.

### Boolean / checkbox input type

The `Boolean` type maps to a checkbox that can be toggled by the user.

When checked, it maps to `true` in the respective field. Additionally, refer to the [FEEL](#adding-feel-editor-support-feel) section to use `Booleans` as expressions.

### Dropdown input type

> The `Dropdown` type allows users to select from a number of pre-defined options that are stored in a custom properties `choices` attribute as `{ name, value }` pairs:

Reviewer: ❓ "custom property `choices` attribute"? Or just "stored in a `choices` attribute"?

The following input types support the `feel` property:

- `String`
- `Text`
- `Number`
- `Boolean`

### FEEL required

The properties panel will display the field as a FEEL editor and will show a visual indication that a FEEL expression is required:

```json
{
  "label": "Required FEEL Expression",
  "type": "String",
  "feel": "required",
  ...
}
```

### FEEL optional

The properties panel will show an indicator to switch to a FEEL expression. When activated, the field display as a FEEL editor:

```json
    {
  "label": "Optional FEEL Expression",
  "type": "String",
  "feel": "optional",
  ...
}
```

For `Boolean` and `Number` fields, the value will always be persisted as a FEEL expression. This ensures that the value will not be interpreted as a string when evaluated in the engine.

### FEEL static

The value of `feel: static` is only valid for `Boolean` and `Number` fields. Similar to [FEEL optional](#feel-optional), the value of the field will be persisted as a FEEL expression. However, there is no toggle to switch to a FEEL editor and ensures only a static value can be entered:

```json
{
  "label": "Static FEEL value",
  "type": "Number",
  "feel": "static",
  ...
}
```

> For binding types `zeebe:input` and `zeebe:output`, `feel: static` is the value used in case of missing `feel` property.

Author: Hah, I never read that sentence before. (I think). Learned something new :D

```json
{
  ...,
  "value": "aTaskType",
  "binding": {
    "type": "zeebe:taskDefinition:type"
  }
}
```

### Extension properties: `zeebe:property`

| **Binding `type`**          | `zeebe:property`                                      |
| --------------------------- | ----------------------------------------------------- |
| **Valid property `type`'s** | `String`<br />`Text`<br />`Hidden`<br />`Dropdown`    |
| **Binding parameters**      | `name`: The name of the property                      |
| **Mapping result**          | `<zeebe:property name="[name]" value="[userInput] />` |

The `zeebe:property` binding allows you to set any arbitrary property for an outside system. It does not impact execution of the Zeebe engine.

```json
{
  ...,
  "value": "{\"outputVar\": 5}",
  "binding": {
    "type": "zeebe:property",
    "name": "camundaModeler:exampleOutputJson"
  }
}
```

### Primitive BPMN properties: `property`

| **Binding `type`**          | `property`                       |
| --------------------------- | -------------------------------- |
| **Valid property `type`'s** | All property types are supported |
| **Binding parameters**      | `name`: The name of the property |

> | **Mapping result**          | `<... [name]=[userInput] ... />` |

Author: On user input, we'd need them

We support optional bindings that do not persist empty values in the underlying BPMN 2.0 XML.

If a user removes the value in the configured control, it will also remove the mapped element.
The following binding types can be `optional`:

- [`zeebe:input`](#input-mapping-zeebeinput)
- [`zeebe:output`](#output-mapping-zeebeoutput)
- [`zeebe:taskHeader`](#header-zeebetaskheader)
- [`zeebe:property`](#extension-properties-zeebeproperty)

Example:

```json
{
  "label": "Request",
  "type": "String",
  "optional": true,
  "binding": {
    "type": "zeebe:input",
    "name": "request"
  }
}
```

## Grouping fields: `group`

Associate a field with a group (ID) via the fields `group` key:

```json
{
  ...,
  "groups": [
    {
      "id": "definition",
      "label": "Task definition",
      "openByDefault": true
    }
  ],
  "properties": [
    {
      "group": "definition",
      "label": "Task type",
      "type": "String",
      "value": "http-job",
      "binding": {
        "type": "zeebe:taskDefinition",
        "property": "type"
      }
    },
    ...
  ]
}
```

## Validating user input: `constraints`

Custom fields may have a number of constraints associated with them:

> - `notEmpty : Boolean`: Input must be non-empty

Reviewer: 🔧 Only if this is set to `true`, right?

- `minLength : Integer`: Minimal length for the input
- `maxLength : Integer`: Maximal length for the input
- `pattern : Object`: Regular expression to match the input against

### Validating against a regex: `pattern`

Set `pattern` to a regular expression to ensure the user's input matches the pattern.

> Together with the `pattern` constraint, you can define your custom error message:

Reviewer: 🔧 "As part of the `pattern` constraint, you can define a custom error message:"? or "Together with the `pattern` value, you can define a custom error message:"? Either way, I'd change "your" to "a".

Author: makes sense

```json
{
  "label": "Web service URL",
  "type": "String",
  "binding": {
    ...
  },
  "constraints": {
    "notEmpty": true,
    "pattern": {
      "value": "https://.*",
      "message": "Must be https URL"
    }
  }
}
```

:::warning

> When a template exposes a property to a user, the template is responsible to show all validation errors in the properties panel.

Reviewer: ```suggestion
When a template exposes a property to a user, the template is responsible for showing all validation errors in the properties panel.
```

> That includes missing compliance with BPMN and Zeebe schema constraints. Thus, you should use `notEmpty` where necessary.

Reviewer: 🔧 I would say, "This includes non-compliance with BPMN and Zeebe schema constraints. You should therefore use `notEmpty` where necessary."

> The Modeler's problems panel will show errors for invalid properties, regardless of a template being applied or not.

Reviewer: ```suggestion
The Modeler's problems panel shows errors for invalid properties, whether or not a template is applied.
```

:::

## Showing properties conditionally: `condition`

> Properties may have a condition which determines when they should be active, depending on the value of another property.

Reviewer: 🔧 I think it would be helpful to have a concrete example here, in one sentence, of when this might be useful.

> When property is **active**, it is displayed in the properties panel, and its value is serialized in the XML.

Reviewer: ```suggestion
When a property is **active**, it is displayed in the properties panel, and its value is serialized in the XML.
```

All the conditions must be met for the property to be active.

There are three possible comparison operators:

- `equals`: Checks if the value is equal to the value defined in the condition.
- `oneOf`: Checks if the value is in the list of values defined in the condition.
- `isActive`: Checks if the referenced property is currently active and not hidden by other conditions.

```json
[
  {
    "id": "httpMethod",
    "label": "HTTP Method",
    "type": "Dropdown",
    "choices": [
      {
        "name": "get",
        "value": "GET"
      },
      {
        "name": "patch",
        "value": "PATCH"
      },
      {
        "name": "post",
        "value": "POST"
      }
    ],
    "binding": {
      ...
    }
  },
  {
    "label": "Request Body",
    "type": "String",
    "binding": {
      ...
    },
    "condition": {
      "allMatch": [
        {
          "property": "httpMethod",
          "oneOf": [
            "patch",
            "post"
          ]
        },
        {
          "property": "...",
          "isActive": "true"
        },
        {
          "property": "...",
          "equals": "someValue"
        }
      ]
    }
  },
  ...
]
```

> ![Display default entries](./img/entries-visible.png)

Reviewer: 🔧 I'm not sure what this screenshot is supposed to illustrate. Maybe it somehow got moved out of place?

## Preventing edits: `editable`

By default, all properties defined in an element template that do not have type `Hidden` are editable.
You can prevent edits by setting the `editable` property to `false`. The property will be displayed in the properties panel but cannot be changed.

```json
{
  "label": "Task type",
  "type": "String",
  "value": "http-job-type",
  "editable": false,
  "binding": {
    "type": "zeebe:taskDefinition",
    "property": "type"
  }
}
```

## Displaying all entries: `entriesVisible`

> Per default, the element template defines the visible entries of the properties panel. All other property controls are hidden.

Reviewer: ```suggestion
By default, the element template defines the visible entries in the properties panel. All other property controls are hidden.
```

> If you want to bring all the default entries back, it is possible to use the `entriesVisible` property.

Reviewer: ```suggestion
To bring all the default entries back, use the `entriesVisible` property.
```

Author: I'll probably add that this is not recommended
